plugins/flatpak/gs-flatpak.c

```diff
@@ -2618,11 +2629,60 @@ gs_flatpak_refine_runtime_for_install (GsFlatpak *self,
 	return TRUE;
 }
 
+static GsApp *
+gs_flatpak_get_services_app_for_runtime (GsPlugin *plugin, GsApp *runtime)
+{
+	const gchar *runtime_id, *runtime_branch, *services_branch, *arch;
+	g_autofree gchar *services_id = NULL;
+	g_autofree gchar *source = NULL;
+
+	runtime_id = gs_app_get_id (runtime);
+	runtime_branch = gs_app_get_branch (runtime);
+
+	if ((g_strcmp0 (runtime_id, "com.endlessm.Platform") == 0 &&
+	     g_strcmp0 (runtime_branch, "eos3.1") == 0) ||
+	    (g_strcmp0 (runtime_id, "com.endlessm.apps.Platform") == 0 &&
+	     g_strcmp0 (runtime_branch, "1") == 0)) {
+		services_id = g_strdup("com.endlessm.EknServices");
+		services_branch = "eos3";
+	} else if (g_strcmp0 (runtime_id, "com.endlessm.apps.Platform") == 0) {
+		services_id = g_strdup_printf("com.endlessm.EknServices%s", runtime_branch);
+		services_branch = "stable";
+	} else {
+		/* Runtime doesn't require an EknServices app */
+		return NULL;
+	}
+
+	/* Construct a GsApp for the EknServices we determined we needed */
+	arch = gs_flatpak_app_get_ref_arch (runtime);
+	source = g_strdup_printf ("app/%s/%s/%s", services_id, arch, services_branch);
+	return gs_flatpak_create_app_from_info (plugin, source, services_id, arch,
```

Comment:
I don't think the app should be created this way. You should rather use `gs_flatpak_create_app` that takes a `FlatpakRef`. That way you also don't have to split up the runtime creation function into yet another app creation one.